imqsauth/audit.go

```diff
@@ -0,0 +1,67 @@
+package imqsauth
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/IMQS/log"
+	"github.com/IMQS/serviceauth"
+)
+
+type IMQSAuditor struct {
+	Url string
+	Log *log.Logger
+}
+
+func NewIMQSAuditor(url string, logger *log.Logger) *IMQSAuditor {
+	return &IMQSAuditor{Url: url, Log: logger}
+}
+
+type Action struct {
+	Who     string        `json:"who"`
+	DidWhat string        `json:"did_what"`
+	ToWhat  string        `json:"to_what"`
+	AtTime  int64         `json:"at_time"`
+	Context ActionContext `json:"context"`
+}
+
+type ActionContext struct {
+	Location string `json:"location"`
+}
+
+func (a *IMQSAuditor) AuditUserAction(identity, clientIp, description string) {
+	action := &Action{Who: identity, DidWhat: description, AtTime: time.Now().Unix()}
+
+	var actionContext ActionContext
+	actionContext.Location = clientIp
+	action.Context = actionContext
+
+	actionStr, err := json.Marshal(action)
+	if err != nil {
+		a.Log.Errorf("Failed to marshal action into json: (%v)", err)
+		return
+	}
+
+	req, err := http.NewRequest("POST", a.Url, bytes.NewBuffer(actionStr))
```

Comment:
This is not great. If you are going to need a buffer as an `io.Reader` anyway, rather use a `json.Encoder` to output your struct to the buffer directly, instead of making a json string first and then writing it to a buffer (in other words, you are currently buffering twice)